### PR TITLE
ACAS_ environment variable overrides Brianbolt/issue928

### LINKED ISSUE #928: ACAS_ environment variables
## Expected Behavior
ACAS should not require a secondary .env file to be configured with environment variables.
ACAS should allow environment variables prefixed with `ACAS_`  to replace configs

### PR DESCRIPTION
**Pull request by brianbolt**
## Description
 - Match `ACAS_*` prefixed environment variables
 - Substitute `_` for `.` in matching environment variables
   - Double `_` (`__`) is an escape character which allows someone to use environment variables for configs with `_` in the name.
   - Match the resulting modified environment variable (case insensitive) to a flat file config and replace its value.

## Related Issue
Fixes #928 

## How Has This Been Tested?
Tested using docker compose and set the following variables and confirmed they overwrote the:

> Docker-compose requires $$ to escape $ so that's why those were used `ACAS_CLIENT_SERVICE_CMPDREG_PERSISTENCE_FULLPATH` but they verify that we have the ability to use environment variables which use other configs just like we used to have.

```
     - ACAS_CLIENT_MODULEMENUS_LOGOTEXT=brian
     - ACAS_SERVER_DATAFILES_RELATIVE__PATH=hamburger
     - ACAS_CLIENT_MODULEMENUS_MENUCONFIGURATIONSETTINGS=[{"isHeader":true,"menuName":"Load Datass","requireUserRoles":["ROLE_ACAS-USERS"],"collapsible":false},{"isHeader":false,"menuName":"Experiment Loader","mainControllerClassName":"GenericDataParserController","autoLaunchName":"generic_data_parser","requireUserRoles":["${"]},{"isHeader":false,"menuName":"Dose Response","mainControllerClassName":"DoseResponseFitWorkflowController","requireUserRoles":["ROLE_ACAS-USERS"]},{"isHeader":false,"menuName":"Protocol Editor","mainControllerClassName":"ProtocolBaseController","autoLaunchName":"protocol_base","requireUserRoles":["ROLE_ACAS-USERS"]},{"isHeader":false,"menuName":"Experiment Editor","mainControllerClassName":"ExperimentBaseController","autoLaunchName":"experiment_base","requireUserRoles":["ROLE_ACAS-USERS"]},{"isHeader":false,"menuName":"Example Thing","mainControllerClassName":"ExampleThingController","autoLaunchName":"example_thing"},{"isHeader":false,"menuName":"Example Thing Browser","mainControllerClassName":"ExampleThingBrowserController","autoLaunchName":"example_thing_browser"},{"isHeader":true,"menuName":"Search and Edit","collapsible":false,"requireUserRoles":["ROLE_ACAS-USERS"]},{"isHeader":false,"menuName":"Protocol Browser","mainControllerClassName":"ProtocolBrowserController","requireUserRoles":["ROLE_ACAS-USERS"]},{"isHeader":false,"menuName":"Experiment Browser","mainControllerClassName":"ExperimentBrowserController","requireUserRoles":["ROLE_ACAS-USERS"]},{"isHeader":true,"menuName":"Admin","collapsible":false,"requireUserRoles":["ROLE_ACAS-ADMINS"]},{"isHeader":false,"menuName":"Project Editor","mainControllerClassName":"ProjectController","autoLaunchName":"project","requireUserRoles":["ROLE_ACAS-ADMINS"]},{"isHeader":false,"menuName":"Project Browser","mainControllerClassName":"ProjectBrowserController","requireUserRoles":["ROLE_ACAS-USERS"]},{"isHeader":false,"menuName":"Author Editor","mainControllerClassName":"AuthorEditorController","autoLaunchName":"author","requireUserRoles":["ROLE_ACAS-ADMINS"]},{"isHeader":false,"menuName":"Author Browser","mainControllerClassName":"AuthorBrowserController","requireUserRoles":["ROLE_ACAS-ADMINS"]},{"isHeader":false,"menuName":"System Test","mainControllerClassName":"SystemTestController","autoLaunchName":"system_test","requireUserRoles":["ROLE_ACAS-ADMINS"]},{"isHeader":false,"menuName":"Label Sequence","mainControllerClassName":"ACASLabelSequenceController","autoLaunchName":"acasLabelSequence","requireUserRoles":["ROLE_ACAS-ADMINS"]},{"isHeader":false,"menuName":"Additional Assay Scientists","mainControllerClassName":"AssayScientistBrowserController","autoLaunchName":"assay_scientist_browser","requireUserRoles":["ROLE_ACAS-ADMINS"]},{"isHeader":true,"menuName":"CmpdReg Admin","requireUserRoles":["ROLE_CMPDREG-ADMINS"],"collapsible":false},{"isHeader":false,"menuName":"CmpdReg Standardization","mainControllerClassName":"StandardizationController","autoLaunchName":"standardization","requireUserRoles":["ROLE_CMPDREG-ADMINS"]},{"isHeader":false,"menuName":"CmpdReg Vendors","mainControllerClassName":"VendorBrowserController","autoLaunchName":"vendor_browser","requireUserRoles":["ROLE_CMPDREG-ADMINS"]},{"isHeader":false,"menuName":"CmpdReg Stereo Categories","mainControllerClassName":"StereoCategoryBrowserController","autoLaunchName":"stereo_category_browser","requireUserRoles":["ROLE_CMPDREG-ADMINS"]},{"isHeader":false,"menuName":"Additional CmpdReg Scientists","mainControllerClassName":"CompoundScientistBrowserController","autoLaunchName":"compound_scientist_browser","requireUserRoles":["ROLE_CMPDREG-ADMINS"]}]
     - ACAS_CLIENT_SERVICE_CMPDREG_PERSISTENCE_FULLPATH=http://$${client.service.cmpdReg.persistence.host}:$${client.service.persistence.port}/$${client.service.cmpdReg.persistence.path}/briantest
```

### CODE REVIEW

File: modules/BuildUtilities/src/server/PrepareConfigFiles.coffee [hunk 1]
```diff
@@ -186,23 +185,40 @@ getProperties = (configDir) =>
...
-		console.info "read env file: #{envFile}"
+	# Read environment variables prefixed with ACAS_ and substitute them in the config file
+	envinmentSubstitutions = {}
+	for key, value of process.env
+		if key.startsWith('ACAS_')
+			# Remove ACAS prefix and lower case
+			newKey = key.substring(5).toLowerCase()
+			# To deal with configs with an underscore, a user can set a double _ inside an environment variable and ACAS will replace it with a single _
+			# Replace double underscores with a unique separator
+			uniqueSeperator = '----------------------------ACAS-UNIQUE-SEPERATOR----------------------------'
```
Comment (bffrost) [inline modules/BuildUtilities/src/server/PrepareConfigFiles.coffee:207]: nit: typo in `seperator` should be `separator`

**Review by bffrost — CHANGES_REQUESTED** [review]
Noted some typos but otherwise this looks great. Very simple.
However I think it's critical that we document this feature somewhere. Could you please add a section to the README explaining the environment variable naming convention and the fact that env vars will override other configs? It would be best with examples both with and without underscores.

File: modules/BuildUtilities/src/server/PrepareConfigFiles.coffee [hunk 2]
```diff
@@ -186,23 +185,40 @@ getProperties = (configDir) =>
...
-	envFiles = envFiles.reverse()
-	envFiles.push "#{acasHome}/conf/.env"
-	console.info "started with process.env and now reading env files in this order (former envs take presidence over latter envs): #{envFiles}"
-	for envFile in envFiles
-		console.info "reading env file: #{envFile}"
-		fileEnv = dotenv.config({ path: envFile} )
-		dotenvexpand(fileEnv)
-		console.info "read env file: #{envFile}"
+	# Read environment variables prefixed with ACAS_ and substitute them in the config file
+	envinmentSubstitutions = {}
```
[the remark this answers is not in the document]
Comment (brianbolt) [reply modules/BuildUtilities/src/server/PrepareConfigFiles.coffee]: Realized I don't even use this variable any more so I removed it.

File: modules/BuildUtilities/src/server/PrepareConfigFiles.coffee (hunk 1 above)
Comment (brianbolt) [reply modules/BuildUtilities/src/server/PrepareConfigFiles.coffee]: Fixed

**Comment by brianbolt** [comment]
I added a section to the readme.

https://github.com/mcneilco/acas/blob/brianbolt/issue928/README.md#configuration

Re-requesting review.

**Review by bffrost — APPROVED** [review]
LGTM. Thanks for adding the README info. I find it clear and easy to follow, and the examples are great.

**Pull request merged**